eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java

```diff
@@ -1998,6 +1998,55 @@ default void appendString(Appendable appendable, String separator)
      */
     <V> Multimap<V, T> groupBy(Function<? super T, ? extends V> function);
 
+    /**
+     * For each element of the iterable, the functions are evaluated and the results of these valuations are collected
+     * into a new map of Multimap, where the first and second functions are evaluated to produce the keys to the map and the inner
+     * Multimap respectively, and the original values are added to the same (or similar) species of collection as the
+     * source iterable.
+     * <p>
+     * Example using a Java 8 method reference:
+     * <pre>
+     * Map&lt;String, Multimap&lt;String, Person&gt;&gt; peopleByCityAndLastName =
+     *     people.groupBy(Person::getCity, Person::getLastName);
+     * </pre>
+     * <p>
+     * Example using an anonymous inner class:
+     * <pre>
+     * Map&lt;String, Multimap&lt;String, Person&gt;&gt; peopleByCityAndLastName =
+     *     people.groupBy(new Function&lt;Person, String&gt;()
+     *     {
+     *         public String valueOf(Person person)
+     *         {
+     *             return person.getCity();
+     *         }
+     *     },
+     *     new Function&lt;Person, String&gt;()
+     *     {
+     *         public String valueOf(Person person)
+     *         {
+     *             return person.getLastName();
+     *         }
+     *     }
+     *     );
+     * </pre>
+     *
+     * @since 9.x
+     */
+    default <V, W> MutableMap<V, Multimap<W, T>> groupBy2(Function<? super T, ? extends V> function1, Function<? super T, ? extends W> function2)
```

Comment:
Since we're returning a type with nested generics, we lose the opportunity to use type variance, so no MutableMultimap on mutable types and ImmutableMultimap on immutable types. This is similar to chunk(). At runtime, they should probably really be mutable or immutable as appropriate. The outer type can still vary normally. So MapIterable here, MutableMap on MutableCollection, ImmutableMap on ImmutableCollection, etc. The overrides should go in the same places where groupBy is already overridden.

Reply:
We've used the type parameters `T` and `V` since before I've been involved in the project. I think `V` must be for `Value` or something. Instead of adding `W`, I'd recommend numbered types like `V1` and `V2` or something similar.